<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>
<body>

<!-- Execute Asynchronous Functions in Parallel -->

<!-- Given an array of asynchronous functions functions, return a new promise promise. Each function in the array accepts no arguments and returns a promise. All the promises should be executed in parallel.

promise resolves:
When all the promises returned from functions were resolved successfully in parallel. The resolved value of promise should be an array of all the resolved values of promises in the same order as they were in the functions. The promise should resolve when all the asynchronous functions in the array have completed execution in parallel.

promise rejects:
When any of the promises returned from functions were rejected. promise should also reject with the reason of the first rejection.
Please solve it without using the built-in Promise.all function. -->

    <script>

    function parallel(functions){
        let arr = [];
       // console.log(arr)
        let completed=0;
       return new Promise((resolve,reject) => {
        
        functions.forEach((fn,i) => {
            fn()
            .then(result => {
              arr[i] = result
              completed++
              
              if(completed === functions.length){
                resolve(arr);
              }
            })
            .catch(reject)
        })
        
        if(functions.length === 0){
            resolve([])
        }

       })
    }

    let fn  = parallel([
        () => new Promise(res => setTimeout(() => res(1))),
        () => new Promise(res => setTimeout(() => res(2))),
        () => new Promise(res => setTimeout(() => res(3))),
        //() => new Promise ((res , rej) => setTimeout(() => rej('error occupied'))) // if one promish reject then all promish reject
    ]);

    fn.then(console.log).catch(console.error)

    </script>
</body>
</html>